### PR TITLE
Libtorch cuda graphs

### LINKED ISSUE #2373: 💡 [REQUEST] - Write a LibTorch CUDA Graph tutorial
### 🚀 Descirbe the improvement or the new tutorial

Write a simple tutorial for CUDA Graphs in LibTorch (C++).

### Existing tutorials on this topic

It can be based on existing MNIST tutorial and distributed MNIST.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Aidyn-A**
Fixes #2373

## Checklist

- [ ] The issue that is being fixed is referred in the description (see above "Fixes #ISSUE_NUMBER")
- [ ] Only one issue is addressed in this pull request
- [ ] Labels from the issue that this PR is fixing are added to this pull request
- [ ] No unnecessary issues are included into this pull request.

### CODE REVIEW

File: advanced_source/cpp_cuda_graphs.rst [hunk 1]
```diff
@@ -0,0 +1,177 @@
+Using CUDA Graphs in PyTorch C++ API
+====================================
+
+NVIDIA’s CUDA Graphs have been a part of CUDA Toolkit library since the
+release of `version 10 <https://developer.nvidia.com/blog/cuda-graphs/>`_.
+They are capable of greatly reducing the CPU overhead increasing the
+performance of applications.
+
+In this tutorial we will be focusing on using CUDA Graphs for `C++
```
Comment (svekars) [inline advanced_source/cpp_cuda_graphs.rst:9]:
```suggestion
In this tutorial, we will be focusing on using CUDA Graphs for `C++
```

File: advanced_source/cpp_cuda_graphs.rst [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+and at the same time simple-to-use tool. In fact, CUDA Graphs are used by default
+in ``torch.compile`` of PyTorch 2.0 to boost the productivity of training and inference.
+
+We would like to demonstrate CUDA Graphs usage on PyTorch’s `MNIST
+example <https://github.com/pytorch/examples/tree/main/cpp/mnist>`_.
+The usage of CUDA Graphs in LibTorch (C++ Frontend) is very similar to its
+`Python counterpart <https://pytorch.org/docs/main/notes/cuda.html#cuda-graphs>`_
+but with some differences in syntax and functionality.
+
+Without further ado, let us get started!
```
Comment (svekars) [inline advanced_source/cpp_cuda_graphs.rst:24]:
```suggestion
Getting Started
----------------
```

File: advanced_source/cpp_cuda_graphs.rst [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+    auto data = batch.data.to(device);
+    auto targets = batch.target.to(device);
+    optimizer.zero_grad();
+    auto output = model.forward(data);
+    auto loss = torch::nll_loss(output, targets);
+    loss.backward();
+    optimizer.step();
+  }
+
+Which are basically a forward pass, backward pass and weight updates.
```
Comment (svekars) [inline advanced_source/cpp_cuda_graphs.rst:41]:
```suggestion
The example above includes a forward pass, a backward pass, and weight updates.
```

File: advanced_source/cpp_cuda_graphs.rst [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+
+In this tutorial we will be applying CUDA Graph on all the compute steps via whole-network
```
Comment (svekars) [inline advanced_source/cpp_cuda_graphs.rst:43]:
```suggestion
In this tutorial, we will be applying CUDA Graph on all the compute steps through the whole-network
```

File: advanced_source/cpp_cuda_graphs.rst [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+graph capture. But before doing so, we need to slightly modify the source code. What we need
+to do is preallocate tensors for reusing them in the main training loop like so:
```
Comment (svekars) [inline advanced_source/cpp_cuda_graphs.rst:45]:
```suggestion
to do is preallocate tensors for reusing them in the main training loop. Here is an example implementation:
```

File: advanced_source/cpp_cuda_graphs.rst [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+      torch::Tensor& output,
+      torch::Tensor& loss) {
+    optimizer.zero_grad();
+    output = model.forward(data);
+    loss = torch::nll_loss(output, targets);
+    loss.backward();
+    optimizer.step();
+  }
+
+PyTorch’s CUDA Graphs API is relying on Stream Capture which in our case would used as following:
```
Comment (svekars) [inline advanced_source/cpp_cuda_graphs.rst:83]:
```suggestion
PyTorch’s CUDA Graphs API is relying on Stream Capture which in our case would be used like this:
```

File: advanced_source/cpp_cuda_graphs.rst [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+
+  at::cuda::CUDAGraph graph;
+  at::cuda::CUDAStream captureStream = at::cuda::getStreamFromPool();
+  at::cuda::setCurrentCUDAStream(captureStream);
+
+  graph.capture_begin();
+  training_step(model, optimizer, data, targets, output, loss);
+  graph.capture_end();
+
+Before actual graph capture it is important to run several warm-up iterations on side stream to
```
Comment (svekars) [inline advanced_source/cpp_cuda_graphs.rst:95]:
```suggestion
Before the actual graph capture, it is important to run several warm-up iterations on side stream to
```

File: advanced_source/cpp_cuda_graphs.rst [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+
+.. code-block:: cpp
+
+  at::cuda::CUDAStream warmupStream = at::cuda::getStreamFromPool();
+  at::cuda::setCurrentCUDAStream(warmupStream);
+  for (int iter = 0; iter < num_warmup_iters; iter++) {
+    training_step(model, optimizer, data, targets, output, loss);
+  }
+
+After successful graph capture we can replace ``training_step(model, optimizer, data, targets, output, loss);``
```
Comment (svekars) [inline advanced_source/cpp_cuda_graphs.rst:107]:
```suggestion
After the successful graph capture, we can replace ``training_step(model, optimizer, data, targets, output, loss);``
```

File: advanced_source/cpp_cuda_graphs.rst [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+call via ``graph.replay();`` do the training step.
```
Comment (svekars) [inline advanced_source/cpp_cuda_graphs.rst:108]: should it be "to do the training step"?

File: advanced_source/cpp_cuda_graphs.rst [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+
+The full source code is available in GitHub.
+
```
Comment (svekars) [inline advanced_source/cpp_cuda_graphs.rst:111]: I feel like this can be a separate section. Maybe "Comparing  Training Results" or something like that?

File: advanced_source/cpp_cuda_graphs.rst [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+  Test set: Average loss: 0.0603 | Accuracy: 0.981
+  Train Epoch: 9 [59584/60000] Loss: 0.0762
+  Test set: Average loss: 0.0547 | Accuracy: 0.983
+  Train Epoch: 10 [59584/60000] Loss: 0.0207
+  Test set: Average loss: 0.0525 | Accuracy: 0.983
+
+  real    0m6.952s
+  user    0m7.048s
+  sys    0m0.619s
+
```
Comment (svekars) [inline advanced_source/cpp_cuda_graphs.rst:171]: This should be go under the "Conclusion" section.

File: advanced_source/cpp_cuda_graphs.rst [hunk 12]
```diff
@@ -0,0 +1,177 @@
...
+As we can see, just by applying a CUDA Graph on the `MNIST example
+<https://github.com/pytorch/examples/tree/main/cpp/mnist>`_ we were able to gain the performance
+by more than 6 times for training. This kind of large perf improvement was achievable due to
```
Comment (svekars) [inline advanced_source/cpp_cuda_graphs.rst:174]:
```suggestion
by more than six times for training. This kind of large performance improvement was achievable due to the
```

File: advanced_source/cpp_cuda_graphs.rst [hunk 13]
```diff
@@ -0,0 +1,177 @@
...
+small model size. In case of larger models with heavy GPU usage the CPU overhead is less impactful
```
Comment (svekars) [inline advanced_source/cpp_cuda_graphs.rst:175]:
```suggestion
small model size. In case of larger models with heavy GPU usage, the CPU overhead is less impactful
```

File: advanced_source/cpp_cuda_graphs.rst [hunk 14]
```diff
@@ -0,0 +1,177 @@
...
+the CUDA Graphs won users’ and developer’s hearts for being a very performant
+and at the same time simple-to-use tool. In fact, CUDA Graphs are used by default
+in ``torch.compile`` of PyTorch 2.0 to boost the productivity of training and inference.
+
+We would like to demonstrate CUDA Graphs usage on PyTorch’s `MNIST
+example <https://github.com/pytorch/examples/tree/main/cpp/mnist>`_.
+The usage of CUDA Graphs in LibTorch (C++ Frontend) is very similar to its
+`Python counterpart <https://pytorch.org/docs/main/notes/cuda.html#cuda-graphs>`_
+but with some differences in syntax and functionality.
+
```
[the remark this answers is not in the document]
Comment (svekars) [reply advanced_source/cpp_cuda_graphs.rst:23]: Also add a note: View and edit this tutorial in github. Like this: https://pytorch.org/tutorials/intermediate/ddp_tutorial.html

File: advanced_source/cpp_cuda_graphs.rst [hunk 15]
```diff
@@ -0,0 +1,177 @@
...
+  at::cuda::CUDAStream warmupStream = at::cuda::getStreamFromPool();
+  at::cuda::setCurrentCUDAStream(warmupStream);
+  for (int iter = 0; iter < num_warmup_iters; iter++) {
+    training_step(model, optimizer, data, targets, output, loss);
+  }
+
+After successful graph capture we can replace ``training_step(model, optimizer, data, targets, output, loss);``
+call via ``graph.replay();`` do the training step.
+
+The full source code is available in GitHub.
```
Comment (svekars) [inline advanced_source/cpp_cuda_graphs.rst:110]:
```suggestion
```
We say that above

**Review by svekars — CHANGES_REQUESTED** [review]

File: advanced_source/cpp_cuda_graphs/mnist.cpp [hunk 16]
```diff
@@ -0,0 +1,378 @@
...
+    torch::Tensor& targets,
+    torch::Tensor& output,
+    torch::Tensor& loss,
+    at::cuda::CUDAGraph& graph,
+    bool use_graph) {
+  model.train();
+
+  size_t batch_idx = 0;
+
+  for (auto& batch : data_loader) {
```
Comment (malfet) [inline advanced_source/cpp_cuda_graphs/mnist.cpp:133]:
```suggestion
  for (const auto& batch : data_loader) {
```

File: advanced_source/cpp_cuda_graphs/mnist.cpp [hunk 17]
```diff
@@ -0,0 +1,378 @@
...
+  at::cuda::CUDAStream captureStream = at::cuda::getStreamFromPool();
+  at::cuda::CUDAStream legacyStream = at::cuda::getCurrentCUDAStream();
+
+  at::cuda::setCurrentCUDAStream(warmupStream);
+  stream_sync(captureStream, legacyStream);
+
+  for (int iter = 0; iter < num_warmup_iters; iter++) {
+    test_step(model, data, targets, output, loss);
+    total_loss = total_loss + loss;
+    total_correct = total_correct + output.argmax(1).eq(targets).sum();
```
Comment (malfet) [inline advanced_source/cpp_cuda_graphs/mnist.cpp:194]: Why not use `+=` here?
```suggestion
    total_loss += loss;
    total_correct += output.argmax(1).eq(targets).sum();
```

File: advanced_source/cpp_cuda_graphs/mnist.cpp [hunk 18]
```diff
@@ -0,0 +1,378 @@
...
+        test_dataset_size,
+        test_data,
+        test_targets,
+        test_output,
+        test_loss,
+        test_total_loss,
+        test_total_correct,
+        test_graph,
+        use_test_graph);
+  }
```
Comment (malfet) [inline advanced_source/cpp_cuda_graphs/mnist.cpp:377]: Add `return 0` and some message indicating that work was successful
```suggestion
  }
  std::cout << " Training/testing complete" << std::endl;
  return 0;
```

**Comment by ezyang** [comment]
deferring to @malfet on this

File: advanced_source/cpp_cuda_graphs.rst [hunk 19]
```diff
@@ -0,0 +1,196 @@
+Using CUDA Graphs in PyTorch C++ API
+====================================
+
+.. note::
+   |edit| View and edit this tutorial in `GitHub <https://github.com/pytorch/tutorials/blob/main/advanced_source/cpp_cuda_graphs.rst>`__.
+
+.. note::
+   The full source code is available on `GitHub <https://github.com/pytorch/tutorials/blob/main/advanced_source/cpp_cuda_graphs>`__.
```
Comment (svekars) [inline advanced_source/cpp_cuda_graphs.rst:8]:
```suggestion
.. note::
   |edit| View and edit this tutorial in `GitHub <https://github.com/pytorch/tutorials/blob/main/advanced_source/cpp_cuda_graphs.rst>`__. The full source code is available on `GitHub <https://github.com/pytorch/tutorials/blob/main/advanced_source/cpp_cuda_graphs>`__.
```
Let's just make it one note

File: advanced_source/cpp_cuda_graphs.rst [hunk 20]
```diff
@@ -0,0 +1,196 @@
...
+
+Prerequisites:
+
+-  `Using the PyTorch C++ Frontend <../advanced_source/cpp_frontend.html>`__
+-  `CUDA semantics <https://pytorch.org/docs/master/notes/cuda.html>`__
+-  `Pytorch 2.0 or later`
+-  `CUDA 11 or later`
```
Comment (svekars) [inline advanced_source/cpp_cuda_graphs.rst:15]:
```suggestion
-  Pytorch 2.0 or later
-  CUDA 11 or later
```

File: advanced_source/cpp_cuda_graphs/CMakeLists.txt [hunk 21]
```diff
@@ -0,0 +1,31 @@
+cmake_minimum_required(VERSION 3.1 FATAL_ERROR)
```
Comment (malfet) [inline advanced_source/cpp_cuda_graphs/CMakeLists.txt:1]: I don't think it will actually compiles with 3.1
```suggestion
cmake_minimum_required(VERSION 3.18 FATAL_ERROR)
```

**Review by malfet — APPROVED** [review]
LGTM, adding perf numbers to PR description
